core/src/history.rs

```diff
@@ -22,18 +25,62 @@ impl ExchangeHistory {
             .map(|(event, batch_id)| (event, BatchId(batch_id as _)))
             .collect();
 
-        Ok(ExchangeHistory { _events: events })
+        Ok(ExchangeHistory::from_events(events))
+    }
+
+    fn from_events(events: Vec<(batch_exchange::Event, BatchId)>) -> Self {
+        // NOTE: Since history methods depend on events being sorted, assert
+        // this invariant is true!
+        debug_assert!(events.windows(2).all(|event_pair| {
+            let (_, previous_event_batch) = event_pair[0];
+            let (_, next_event_batch) = event_pair[1];
+            previous_event_batch <= next_event_batch
+        }));
+
+        ExchangeHistory { events }
     }
 
     /// Reads historic exchange events from an `Orderbook` filestore.
     pub fn from_filestore(orderbook_filestore: impl AsRef<Path>) -> Result<Self> {
         ExchangeHistory::read(File::open(orderbook_filestore)?)
     }
+
+    /// Returns an iterator of all events, including their batch IDs, that were
+    /// emitted until the end of the specified batch.
+    fn events_until_end_of_batch(
+        &self,
+        batch_id: impl Into<BatchId>,
+    ) -> impl Iterator<Item = (&batch_exchange::Event, BatchId)> + '_ {
+        let end_batch_id = batch_id.into();
+        self.events
+            .iter()
+            .take_while(move |(_, batch_id)| *batch_id <= end_batch_id)
+            .map(|(event, batch_id)| (event, *batch_id))
+    }
+
+    /// Returns the finalized orderbook at a historic batch.
+    ///
+    /// The finalized orderbook includes all account state changes and orders
+    /// that are consisdered by the smart contract for solving the subsequent
+    /// batch.
+    pub fn auction_data_for_batch(
+        &self,
+        batch_id: impl Into<BatchId>,
+    ) -> Result<(AccountState, Vec<Order>)> {
+        let batch_id = batch_id.into();
+
+        State::from_events(
+            self.events_until_end_of_batch(batch_id)
+                .map(|(event, batch_id)| (event, batch_id.into())),
+        )?
+        .normalized_auction_state_at_beginning_of_batch(batch_id.next().into())
```

Comment:
Probably fine but for some historical analysis we might not want to normalize as it removes some information.

Reply:
I'm OK either way. My main reasoning for doing this is:
- AFAIK, In the SC there is no real difference between an untouched balance (either a user that never used the exchange or never used a specific token) and a 0 balance (deposit + withdraw request + withdraw)
- This is not true for orders, but we already lose information about cancelled and deleted orders, so I thought it would be easier to use some normalized view for orders to remove any ambiguity.

Reply:
What information do we lose? Is this about orders that don't have sufficient balance or are not valid? I agree that for now we can normalize it away.

Technically though, this could be an issue since we can receive proceeds in a trade that can be used in the same settlement for selling a token (e.g. if I am willing to sell and buy DAI but have no DAI at the moment I could be getting it in a first trade from token A and sell it in the same batch for token B). However, the price estimator does not cope with this today so this information is not needed.

Reply:
> Technically though, this could be an issue since we can receive proceeds in a trade that can be used in the same settlement for selling a token

Thats a really good point! I I would be curious to know if the solver considers these orders ATM (@marcovc or @twalth3r can you confirm confirm?).

Reply:
The current situation is to forbid it - orders max sell amount are capped by the account balances. That means that a person cannot sell tokens if their account balance is zero. However, there is an ongoing discussion to lift this constraint.

Reply:
I just realized,
```
        .filter(|order| order.remaining_sell_amount > 0)
```

We only filter out orders with 0 remaining sell amount, but orders with remaining sell amount and 0 balance are still left in. So this shouldn't be a concern :smile:

Reply:
I had to dig into the code to figure out what it meant to "normalize_auction_data" and found that this is more like a reduce or filter auction data that simply removed irrelevant details. More precisely:

```
/// Removes empty orders and token balances for which there is
/// not at least one sell order by that user
```

I would like to propose that maybe we use a term like `reduce_auction_data` since normalization usually refers to factoring out a scalar from the whole. Of course, this would be for a separate PR, but I this confusing term took me down a rabbit hole.

Reply:
I also find the term normalize a bit vague here (I believe it's more based on https://en.wikipedia.org/wiki/Database_normalization than scalar extraction). I'd be very happy about a rename.

Reply:
I think this should be `canonicalize`, since we are creating a canonical representation of an orderbook.

Reply:
Yes. I’d be much happier with this term here.

Reply:
This was implemented as part of https://github.com/gnosis/dex-services/pull/1063 :smile: